src/cogent3/core/new_alphabet.py

```diff
@@ -0,0 +1,255 @@
+    elif num_elements < 2**16:
+        dtype = numpy.uint16
+    elif num_elements < 2**32:
+        dtype = numpy.uint32
+    elif num_elements < 2**64:
+        dtype = numpy.uint64
+    else:
+        raise NotImplementedError(f"{num_elements} is too big for 64-bit integer")
+
+    return dtype
+
+
+def consistent_words(
+    words: typing.Sequence[typing.Union[str, int]], length: OptInt = None
+) -> None:
+    """make sure all alphabet elements are unique and have the same length"""
+    if not words:
+        raise ValueError("no words provided")
+
+    if len(set(words)) != len(words):
+        raise ValueError(f"duplicate elements in {words}")
+
+    lengths = {1} if isinstance(words[0], int) else {len(w) for w in words}
+    if len(lengths) != 1:
+        raise ValueError(f"mixed lengths {lengths} in {words}")
+    l = list(lengths)[0]
+    if length and l != length:
+        raise ValueError(f"word length {l} does not match expected {length}")
+
+
+class bytes_to_array:
+    """converts utf8 to numpy array"""
+
+    def __init__(self, chars: bytes, dtype):
+        # we want a bytes translation map
+        self._table = b"".maketrans(
+            chars,
+            bytes(bytearray(range(len(chars)))),
+        )
+        self.dtype = dtype
+
+    def __call__(self, seq: bytes) -> numpy.ndarray:
+        b = seq.translate(self._table)
+        return numpy.array(memoryview(b), dtype=self.dtype)
+
+
+class array_to_bytes:
+    """converts utf8 to numpy array"""
+
+    def __init__(self, chars, dtype):
+        # we want a bytes translation map
+        self._table = b"".maketrans(
+            bytes(bytearray(range(len(chars)))),
+            chars,
+        )
+        self.dtype = dtype
+
+    def __call__(self, seq: numpy.ndarray) -> bytes:
+        b = seq.tobytes().translate(self._table)
+        return bytearray(b)
+
+
+class CharAlphabet(tuple, AlphabetABC, MonomerAlphabetABC):
+    def __new__(cls, chars: typing.Sequence[StrORBytes], gap: OptStr = None):
+        if not chars:
+            raise ValueError(f"cannot create empty {cls.__name__!r}")
+
+        if gap is not None:
+            assert _coerce_to_type(chars[0], gap) in chars
+
+        consistent_words(chars, length=1)
+        return tuple.__new__(cls, chars, gap=gap)
+
+    def __init__(self, chars: typing.Sequence[StrORBytes], gap: str = "-"):
+        self._gap = gap
+        self.dtype = get_array_type(len(self))
+        self._chars = set(self)  # for quick lookup
+        try:
+            byte_chars = b"".join([c.encode("utf8") for c in chars])
+        except AttributeError:
+            byte_chars = bytes(bytearray(chars))
+        self._bytes2arr = bytes_to_array(byte_chars, dtype=self.dtype)
+        self._arr2bytes = array_to_bytes(byte_chars, dtype=self.dtype)
+
+    @property
+    def motif_len(self) -> int:
+        # this is enforced by consistent_words(length=1) in the constructor
+        return 1
+
+    @functools.singledispatchmethod
```

Comment:
@GavinHuttley you need to have a dispatch for when it is already a `numpy.ndarray` where it just returns it